File: crates/xtask/src/method_census.rs
//! `cargo run -p xtask -- method-census [--check]`: records what CRuby's whole
//! reachable module tree owns, into `conformance/method-census.tsv`.
//!
//! The dump is checked in so `crates/zeo/tests/method_census.rs` can gate
//! coverage on a machine with no ruby -- that test compiles the SAME walker
//! (`tools/method_census.rb`) through zeo and diffs the two. Regenerating is an
//! explicit act; `--check` re-runs the oracle and fails if the committed copy
//! is stale, for a CI job that does have the pinned ruby.

use std::path::{Path, PathBuf};
use std::process::{Command, ExitCode};

const DUMP: &str = "conformance/method-census.tsv";
const WALKER: &str = "tools/method_census.rb";

pub fn main(root: &Path, args: &[String]) -> ExitCode {
    let check = args.iter().any(|a| a == "--check");

    let ruby = resolve_ruby(root);
    let out = match Command::new(&ruby)
        .arg("--disable-gems")
        .arg(root.join(WALKER))
        .current_dir(root)
        .output()
    {
        Ok(o) if o.status.success() => o,
        Ok(o) => {
            eprintln!("the ruby oracle exited with {}", o.status);
            eprint!("{}", String::from_utf8_lossy(&o.stderr));
            return ExitCode::FAILURE;
        }
        Err(e) => {
            eprintln!("cannot run the ruby oracle at {}: {e}", ruby.display());
            return ExitCode::FAILURE;
        }
    };
    let dump = String::from_utf8_lossy(&out.stdout).into_owned();

    let target = root.join(DUMP);
    if check {
        let committed = std::fs::read_to_string(&target).unwrap_or_default();
        if committed == dump {
            println!("{DUMP} is up to date");
            return ExitCode::SUCCESS;
        }
        eprintln!("{DUMP} is stale -- re-run `cargo run -p xtask -- method-census`");
        return ExitCode::FAILURE;
    }

    if let Some(parent) = target.parent() {
        let _ = std::fs::create_dir_all(parent);
    }
    if let Err(e) = std::fs::write(&target, &dump) {
        eprintln!("cannot write {}: {e}", target.display());
        return ExitCode::FAILURE;
    }
    let modules = dump.lines().filter(|l| l.contains("\ti\t")).count();
    println!(
        "wrote {DUMP} ({modules} modules, {} lines)",
        dump.lines().count()
    );
    ExitCode::SUCCESS
}

/// `mise which ruby` (falling back to bare `ruby`), the same resolution the
/// golden harness uses, so both come from the `mise.toml`-pinned oracle.
fn resolve_ruby(cwd: &Path) -> PathBuf {
    if let Ok(out) = Command::new("mise")
        .arg("which")
        .arg("ruby")
        .current_dir(cwd)
        .output()
        && out.status.success()
    {
        let path = String::from_utf8_lossy(&out.stdout).trim().to_owned();
        if !path.is_empty() {
            return PathBuf::from(path);
        }
    }
    PathBuf::from("ruby")
}

File: tools/lib/zeo_dev/commands/method_census.rb
# frozen_string_literal: true

module ZeoDev
  module Commands
    # Records what CRuby's whole reachable module tree owns, into
    # `conformance/method-census.tsv`.
    #
    # The dump is checked in so `crates/zeo-tests/tests/method_census.rs` can
    # gate coverage on a machine with no ruby -- that test compiles the SAME
    # walker (`tools/method_census.rb`) through zeo and diffs the two.
    # Regenerating is an explicit act; `--check` re-runs the oracle and fails
    # if the committed copy is stale, for a CI job that does have the pinned
    # ruby.
    class MethodCensus < Cli
      DUMP = "conformance/method-census.tsv"
      WALKER = "tools/method_census.rb"

      def self.summary = "record conformance/method-census.tsv from the oracle"
      def self.banner = "usage: zeo-dev method-census [--check]"

      def defaults = { check: false }

      def options(o)
        o.on("--check", "fail if the committed copy is stale, and write nothing") do
          opts[:check] = true
        end
      end

      def run
        expect_no_args!
        # `--disable-gems` and nothing else: the walker reads the core tree,
        # and any flag added here changes the recorded bytes.
        res = Exec.run([Ruby.oracle, "--disable-gems", File.join(ROOT, WALKER)], chdir: ROOT,
                                                                                capture_stdout: true)
        unless res.success?
          warn "the ruby oracle exited with #{res.code.inspect}"
          warn res.stderr
          return 1
        end
        dump = res.stdout
        target = File.join(ROOT, DUMP)

        return Tsv.check(target, dump, label: DUMP) if opts[:check]

        Tsv.write(target, dump)
        modules = dump.lines.count { |l| l.include?("\ti\t") }
        puts "wrote #{DUMP} (#{modules} modules, #{dump.lines.count} lines)"
        0
      end
    end
  end
end
